File: src/main/java/com/teamwizardry/librarianlib/api/gui/components/mixin/DragMixin.java
package com.teamwizardry.librarianlib.api.gui.components.mixin;

import java.util.function.Function;

import com.teamwizardry.librarianlib.api.gui.GuiComponent;
import com.teamwizardry.librarianlib.api.gui.HandlerList;
import com.teamwizardry.librarianlib.math.Vec2;

public class DragMixin<T extends GuiComponent<?>> {
	
	protected T component;
	protected Function<Vec2, Vec2> constraints;
	
	public boolean mouseDown = false;
	public Vec2 clickPos = Vec2.ZERO;
	
	public final HandlerList<IDragEvent<T>> pickup = new HandlerList<>();
	public final HandlerList<IDragEvent<T>> drop = new HandlerList<>();
	public final HandlerList<IDragEvent<T>> drag = new HandlerList<>();
	
	public DragMixin(T component, Function<Vec2, Vec2> constraints) {
		this.component = component;
		this.constraints = constraints;
		init();
	}
	
	private void init() {
		component.mouseDown.add( (c, pos, button) -> {
			if(!mouseDown && c.isMouseOver(pos)) {
				mouseDown = true;
				clickPos = pos;
				pickup.fireAll((h) -> h.handle(component, pos));
				return true;
			}
			return false;
		});
		component.mouseUp.add( (c, pos, button) -> {
			if(mouseDown)
				drop.fireAll((h) -> h.handle(component, pos));
			mouseDown = false;
			return false;
		});
		component.preDraw.add( (c, pos, partialTicks) -> {
			if(mouseDown) {
				Vec2 newPos = constraints.apply(  c.getPos().add(pos).sub(clickPos)  );
				
				if(!newPos.equals(c.getPos())) {
					c.setPos(newPos);
					drag.fireAll((h) -> h.handle(component, newPos));
				}
			}
		});
	}
	
	@FunctionalInterface
	public static interface IDragEvent<T> {
		void handle(T component, Vec2 pos);
	}
}

File: src/main/java/com/teamwizardry/librarianlib/client/gui/mixin/DragMixin.kt
package com.teamwizardry.librarianlib.client.gui.mixin

import com.teamwizardry.librarianlib.client.gui.EnumMouseButton
import com.teamwizardry.librarianlib.client.gui.GuiComponent
import com.teamwizardry.librarianlib.common.util.event.Event
import com.teamwizardry.librarianlib.common.util.event.EventCancelable
import com.teamwizardry.librarianlib.common.util.math.Vec2d

class DragMixin<T : GuiComponent<T>>(protected var component: T, protected var constraints: (Vec2d) -> Vec2d) {

    class DragPickupEvent<T : GuiComponent<*>>(val component: T, val mousePos: Vec2d, val button: EnumMouseButton) : EventCancelable()
    class DragDropEvent<T : GuiComponent<*>>(val component: T, val mousePos: Vec2d, val button: EnumMouseButton) : EventCancelable()
    class DragMoveEvent<T : GuiComponent<*>>(val component: T, val mousePos: Vec2d, val clickedPoint: Vec2d, val pos: Vec2d, var newPos: Vec2d, val button: EnumMouseButton) : Event()

    var mouseDown: EnumMouseButton? = null
    var clickPos = Vec2d.ZERO

    init {
        init()
    }

    private fun init() {
        component.BUS.hook(GuiComponent.MouseDownEvent::class.java) { event ->
            if (mouseDown == null && event.component.mouseOver && !component.BUS.fire(DragPickupEvent(event.component, event.mousePos, event.button)).isCanceled()) {
                mouseDown = event.button
                clickPos = event.mousePos
                event.cancel()
            }
        }
        component.BUS.hook(GuiComponent.MouseUpEvent::class.java) { event ->
            if (mouseDown == event.button && !component.BUS.fire(DragDropEvent(event.component, event.mousePos, event.button)).isCanceled()) {
                mouseDown = null
                event.cancel()
            }
        }

        component.BUS.hook(GuiComponent.PreDrawEvent::class.java) { event ->
            val mouseButton = mouseDown
            if (mouseButton != null) {
                val newPos = constraints(event.component.pos.add(event.mousePos).sub(clickPos))

                if (newPos != event.component.pos) {
                    event.component.pos = event.component.BUS.fire(
                            DragMoveEvent(event.component, event.mousePos, clickPos, event.component.pos, newPos, mouseButton)
                    ).newPos
                }
            }
        }
    }
}
